File: Spring/EventViewController.m
//
//  EventViewController.m
//  Spring
//
//  Created by Joseph Asperger on 11/8/14.
//
//

#import "EventViewController.h"
#import "EventHeaderCell.h"
#import "EventInterestCell.h"
#import "EventWritePostCell.h"
#import "EventPostCell.h"

#define WRITE_POST_SECTION 2
#define DISPLAY_POSTS_SECTION_START 3

@interface EventViewController ()

@property UILabel *interestLabel;
@property UIButton *joinButton;
@property UITextField *postField;
@property BOOL isWritingPost;
@property NSArray *posts;
@property BOOL didLoadPosts;

@end

@implementation EventViewController

-(void) viewDidLoad{
    self.navigationItem.title = self.event[@"displayName"];
    //UITapGestureRecognizer *gestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(hideKeyboard)];
    //[self.tableView addGestureRecognizer:gestureRecognizer];
    [self loadPosts];
}

- (void) loadPosts{
    PFQuery *postsQuery = [PFQuery queryWithClassName:@"EventPost"];
    [postsQuery whereKey:@"event" equalTo:self.event];
    [postsQuery includeKey:@"user"];
    [postsQuery orderByDescending:@"createdAt"];
    [postsQuery findObjectsInBackgroundWithBlock:^(NSArray *objects, NSError *error) {
        if (!error){
            self.posts = objects;
            self.didLoadPosts = YES;
            [self.tableView reloadData];
        }
    }];
}

-(void) joinButtonPressed{
    [self.joinButton setImage:[UIImage imageNamed:@"Checkmark"] forState:UIControlStateNormal];
    [self.joinButton setTitle:@"Joined" forState:UIControlStateNormal];
    self.joinButton.userInteractionEnabled = NO;
}

- (void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.section == WRITE_POST_SECTION){
        if (!self.isWritingPost){
            self.isWritingPost = YES;
            [tableView reloadRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
        }
    }
}

-(void) hideKeyboard{
    [self.postField resignFirstResponder];
}

-(void)keyboardWillHide:(NSNotification *)notification
{
    [self.view setFrame:CGRectMake(0,0,320,460)];
}

- (NSInteger) numberOfSectionsInTableView:(UITableView *)tableView{
    if (!self.didLoadPosts){
        return 3;
    }
    else{
        return 3 + [self.posts count];
    }
    
}

- (NSInteger) tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return 1;

}

- (CGFloat) tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.section == 0){
        return 163;
    }
    else if (indexPath.section == 1){
        return 44;
    }
    else if (indexPath.section == WRITE_POST_SECTION){
        if (self.isWritingPost){
            return 70;
        }else{
            return 44;
        }
    }
    else{
        return 60;
    }
}

-(UITableViewCell*) tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.section == 0){
        static NSString *simpleTableIdentifier = @"EventHeaderCell";
        
        EventHeaderCell *cell = [tableView dequeueReusableCellWithIdentifier:simpleTableIdentifier forIndexPath:indexPath];
        
        PFUser *creator = self.event[@"creator"];
        PFObject *activity = self.event[@"activity"];
        
        cell.nameLabel.text = self.event[@"name"];
        [cell.nameLabel sizeToFit];
        NSDate *date = self.event[@"time"];
        NSDateFormatter *dateFormatter = [NSDateFormatter new];
        [dateFormatter setDateStyle:NSDateFormatterMediumStyle];
        [dateFormatter setTimeStyle:NSDateFormatterShortStyle];
        cell.timeLabel.text = [dateFormatter stringFromDate:date];
        cell.creatorNameLabel.text = creator[@"displayName"];
        [cell.activityButton setTitle:activity[@"name"] forState:UIControlStateNormal];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        return cell;
    }
    else if (indexPath.section == 1){
        static NSString *simpleTableIdentifier = @"EventInterestCell";
        
        EventInterestCell *cell = [tableView dequeueReusableCellWithIdentifier:simpleTableIdentifier forIndexPath:indexPath];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        self.joinButton = cell.joinButton;
        self.interestLabel = cell.interestLabel;
        self.interestLabel.text = [NSString stringWithFormat:@"%ld People Interested", [self.event[@"numberInterested"] integerValue]];
        [self.joinButton addTarget:self action:@selector(joinButtonPressed) forControlEvents:UIControlEventTouchUpInside];
        return cell;
    }
    else if (indexPath.section == WRITE_POST_SECTION){
        if (!self.isWritingPost){
            static NSString *simpleTableIdentifier = @"EventPostButtonCell";
            UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:simpleTableIdentifier forIndexPath:indexPath];
            cell.selectionStyle = UITableViewCellSelectionStyleDefault;
            return cell;
        }
        else{
            EventWritePostCell *cell = [tableView dequeueReusableCellWithIdentifier:@"EventWritePostCell" forIndexPath:indexPath];
            self.postField = cell.postField;
            self.postField.delegate = self;
            [self.postField becomeFirstResponder];
            return cell;
        }
    }
    else if (indexPath.section >= DISPLAY_POSTS_SECTION_START){
        EventPostCell *cell = [tableView dequeueReusableCellWithIdentifier:@"EventPostCell" forIndexPath:indexPath];
        PFObject *post = self.posts[indexPath.section - DISPLAY_POSTS_SECTION_START];
        cell.postLabel.text = post[@"content"];
        return cell;
    }
    else{
        return nil;
    }
}

-(NSString*) tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section{
    if (section >= DISPLAY_POSTS_SECTION_START){
        PFObject *post = self.posts[section - DISPLAY_POSTS_SECTION_START];
        PFUser *user = post[@"user"];
        return user[@"displayName"];
        return nil;
    }
    else{
        return nil;
    }
}

-(void) textFieldDidChange:(UITextField*)textField{

}

-(BOOL)textFieldShouldReturn:(UITextField *)textField
{
    PFObject *eventPost = [PFObject objectWithClassName:@"EventPost"];
    eventPost[@"user"] = [PFUser currentUser];
    eventPost[@"event"] = self.event;
    eventPost[@"content"] = textField.text;
    self.isWritingPost = NO;
    [textField resignFirstResponder];
    [self.tableView reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:0 inSection:WRITE_POST_SECTION]] withRowAnimation:UITableViewRowAnimationAutomatic];
    [eventPost saveInBackgroundWithBlock:^(BOOL succeeded, NSError *error) {
        if (succeeded){
            [self loadPosts];
        }
    }];
    [textField setText:@""];
    return NO;
}


@end

File: Spring/EventViewController.swift
//
//  EventViewController.swift
//  Spring
//
//  Created by Joseph Asperger on 1/18/15.
//
//

import UIKit

class EventViewController: UITableViewController, UITextFieldDelegate {
    
    let headerSection = 0
    let interestSection = 1
    let writePostSection = 2
    let displayPostsSection = 3
    
    var event: PFObject!
    var posts = []
    var didLoadPosts = false
    var isWritingPost = false
    
    var interestLabel: UILabel!
    var joinButton: UIButton!
    var postField: UITextField!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        navigationItem.title = event["displayName"] as? String
        loadPosts()
    }
    
    func joinButtonPressed() {
        joinButton.setImage(UIImage(named: "Checkmark"), forState: UIControlState.Normal)
        joinButton.setTitle("Joined", forState: UIControlState.Normal)
        joinButton.userInteractionEnabled = false
    }
    
    func loadPosts() {
        var postsQuery = PFQuery(className: "EventPost")
        postsQuery.whereKey("event", equalTo: self.event)
        postsQuery.includeKey("user")
        postsQuery.orderByDescending("createdAt")
        postsQuery.findObjectsInBackgroundWithBlock { (objects, error) -> Void in
            if (error == nil){
                self.posts = objects
                self.didLoadPosts = true
                self.tableView.reloadData()
            }
        }
    }
    
    func submitPost(postText: String){
        var eventPost = PFObject(className: "EventPost")
        eventPost["user"] = PFUser.currentUser()
        eventPost["event"] = event
        eventPost["content"] = postText
        eventPost.saveInBackgroundWithBlock { (succeeded, error) -> Void in
            if (succeeded){
                self.loadPosts()
            }
        }
    }
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        submitPost(textField.text)
        self.isWritingPost = false
        textField.resignFirstResponder()
        textField.text = ""
        tableView.reloadRowsAtIndexPaths([NSIndexPath(forRow: 0, inSection: writePostSection)], withRowAnimation: UITableViewRowAnimation.Automatic)
        return false
    }
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        if (indexPath.section == writePostSection){
            if (!isWritingPost){
                isWritingPost = true
                tableView.reloadRowsAtIndexPaths([indexPath], withRowAnimation: UITableViewRowAnimation.Automatic)
            }
        }
    }
    
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        if (didLoadPosts){
            return 4
        }
        else{
            return 3
        }
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if (section == displayPostsSection){
            return posts.count
        }
        return 1
    }
    
    override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        if (indexPath.section == headerSection){
            return 163
        }
        else if (indexPath.section == interestSection){
            return 44
        }
        else if (indexPath.section == writePostSection){
            if (self.isWritingPost){
                return 70
            }
            else{
                return 44
            }
        }
        else{
            return 82 // TODO: implement size to fit
        }
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        if (indexPath.section == headerSection){
            var cell = tableView.dequeueReusableCellWithIdentifier("EventHeaderCell", forIndexPath: indexPath) as EventHeaderCell
            var creator = event["creator"] as PFUser
            var activity = event["activity"] as PFObject
            var date = self.event["time"] as NSDate
            var dateFormatter = NSDateFormatter()
            dateFormatter.dateStyle = NSDateFormatterStyle.MediumStyle
            dateFormatter.timeStyle = NSDateFormatterStyle.ShortStyle
            cell.timeLabel.text = dateFormatter.stringFromDate(date)
            cell.creatorNameLabel.text = creator["displayName"] as? String
            cell.activityButton.setTitle(activity["name"] as? String, forState: UIControlState.Normal)
            return cell
        }
        else if (indexPath.section == interestSection){
            var cell = tableView.dequeueReusableCellWithIdentifier("EventInterestCell", forIndexPath: indexPath) as EventInterestCell
            joinButton = cell.joinButton
            interestLabel = cell.interestLabel
            interestLabel.text = String(format:"%d People Interested", event["numberInterested"] as Int)
            self.joinButton.addTarget(self, action: "joinButtonPressed", forControlEvents: UIControlEvents.TouchUpInside)
            return cell
        }
        else if (indexPath.section == writePostSection){
            if (!self.isWritingPost){
                var cell = tableView.dequeueReusableCellWithIdentifier("EventPostButtonCell", forIndexPath: indexPath) as UITableViewCell
                cell.selectionStyle = UITableViewCellSelectionStyle.Default
                return cell
            }
            else{
                var cell = tableView.dequeueReusableCellWithIdentifier("EventWritePostCell", forIndexPath: indexPath) as EventWritePostCell
                postField = cell.postField
                postField.delegate = self
                postField.becomeFirstResponder()
                return cell
            }
        }
        else{ // is display post cell
            var cell = tableView.dequeueReusableCellWithIdentifier("EventPostCell", forIndexPath: indexPath) as EventPostCell
            var post = posts[indexPath.row] as PFObject
            cell.postLabel.text = post["content"] as? String
            return cell
        }
    }
}

class EventPostCell: UITableViewCell {
    
    @IBOutlet weak var postLabel: UILabel!
}

class EventWritePostCell: UITableViewCell {
    
    @IBOutlet weak var postField: UITextField!
    
}

class EventHeaderCell: UITableViewCell {
    
    @IBOutlet weak var nameLabel: UILabel!
    @IBOutlet weak var creatorNameLabel: UILabel!
    @IBOutlet weak var activityButton: UIButton!
    @IBOutlet weak var timeLabel: UILabel!
    
}

class EventInterestCell: UITableViewCell {
    
    @IBOutlet weak var joinButton: UIButton!
    @IBOutlet weak var interestLabel: UILabel!
    
}


